Use nibabel for image operation

TomMaullin opened the pull request:
This PR addresses the following small miscellaneous image issues:

#20 - The empty mask image error. This should be resolved by switching functions from using FSLstats via the commandline to nilearn.

#14 - Missing slice images in FSL if an image is not stored in the NIDM-Results pack. This should be resolved with a few if statements.

#13 - This should be resolved as #20 is resolved.

#12 - Use the SPM template for SPM slice images and FSL template for displaying FSL slice images.

TomMaullin committed:
clusterp values added

TomMaullin committed:
resolved extent threshold issues

TomMaullin committed:
master updated

TomMaullin committed:
wip

pep8speaks commented:
Hello @TomMaullin! Thanks for updating the PR.









Cheers ! There are no PEP8 issues in this Pull Request. :beers: 

##### Comment last updated on April 03, 2018 at 16:54 Hours UTC

TomMaullin committed:
wip now using nibabel

TomMaullin committed:
wip

TomMaullin committed:
wip now uses SPM template

TomMaullin committed:
wip attempted travis compatibility

TomMaullin committed:
setup adjusted

TomMaullin committed:
manifest update

TomMaullin committed:
setup update

TomMaullin committed:
travis test

TomMaullin committed:
init added

TomMaullin committed:
getVal now uses nilearn

TomMaullin committed:
Merge remote-tracking branch 'origin' into nilearnImage

TomMaullin committed:
scalefactor fix

TomMaullin committed:
PEP8 resolutions

TomMaullin committed:
PEP8 resolutions

TomMaullin committed:
updated comments

TomMaullin committed:
pep8 fix

TomMaullin committed:
now works if fsl slice image is deleted

TomMaullin committed:
now works for missing fsl also

TomMaullin committed:
pep8

TomMaullin committed:
pep8

TomMaullin commented:
This PR is now ready for review. The one issue left with this PR is that the tests written on the testing PR (#22 ) may fail on this update as they are based on the slice image code this aims to replace. I will resolve this once PR #22 is merged.

cmaumet commented:
@TomMaullin thanks for this update! I had a quick look and I can see the code is now using nibabel in many places but do you think it would be feasible to go even further and use nibabel's resampling function rather than manually computing the transformation matrix and then applying it with `flirt`? Here is a snippet of code to do the resampling given:
 - `<PATH_TO_TEMPLATE>`: path to the input template image
 - `<PATH_TO_EXC_SET>`: path to the input excursion set image
 - `<PATH_TO_RESLICED_EXC_SET>`: path to the output resliced excursion set image

```
template_img = nib.load(<PATH_TO_TEMPLATE>)
excset_img = nib.load(<PATH_TO_EXC_SET>
excset_img_resl = resample_from_to(excset_img, template_img)
nib.save(excset_img_resl, <PATH_TO_RESLICED_EXC_SET>)
```

This would make the code much shorter and therefore easier to maintain. Does that sound reasonable?

TomMaullin commented:
Hi @cmaumet ,

Ah yes I'll look into this and let you know how I get on with it!

TomMaullin committed:
merge with master to resolve conflicts

TomMaullin committed:
now use nilearn for resize

TomMaullin committed:
requirements added

TomMaullin commented:
Hi @cmaumet , 

I can confirm the code for resizing images now uses and works with nibabel! Please let me know if you have any further feedback!

TomMaullin committed:
PEP8

cmaumet commented on requirements.txt:
```diff
@@ -1,4 +1,7 @@
 rdflib
 dominate
+scipy
```
Do we need nilearn, sklearn and scipy?

TomMaullin replied:
Scipy is needed as it is a dependency for some of the other modules and is not automatically installed when `setup.py` is run. The others are not, and have been removed.

TomMaullin committed:
Merge branch 'master' of https://github.com/incf-nidash/nidmresults-fslhtml into nilearnImage

TomMaullin committed:
tests updated to match spm template slice images

TomMaullin committed:
test of whether requirements needed

TomMaullin committed:
scipy requirement definately needed

TomMaullin commented:
Hi @cmaumet ,

All feedback on this PR has been addressed and the tests have been updated to account for the change from FSL template to SPM template for SPM-generated axial slice images.

Please let me know if you have any further feedback and/or whether this can now be merged!

cmaumet commented on MANIFEST.in:
```diff
@@ -1 +1,2 @@
-include queries/*.txt
\ No newline at end of file
+include queries/*.txt
+include templates/*.nii
```
@TomMaullin: Usually I would not recommend having NIfTI files in a GitHub repository (because binary files are not handled well). But, because those files will likely not change much in the future, it could be okay to have them... 

 1. Can you check how much space the templates folder is taking up? (I think GitHub repo should be no more than 1GB, including all versions).
 2. Can you check under which licence those NIfTI template images are distributed?

TomMaullin replied:
Yes, I had the same worry, currently they take up 2.7Mb but the T1 can be removed if necessary. I used the T1_skullstripped version for the viewer but created it myself using the T1. 

The T1 image was taken from the OldNorm toolbox in SPM so I guess this means it is under the SPM GNU public license? Do you think this is okay?

cmaumet replied:
Yes, please remove as much as you can to have a minimal number of NIfTI images in the repo.

As far as I know, GPL can be used to licence code but not data. I checked [spm_templates.man](https://github.com/spm/spm12/blob/master/spm_templates.man) but could not find info on licenses for the templates... 

One option: could you automatically download the template (when needed) if it's not available locally, rather than embed the image in the repo? E.g. from https://github.com/spm/spm12

TomMaullin replied:
The issue I would have with automatically downloading from a public repo is that the template I was using is the one I had manually skull-stripped (as the one with the skull included is cropped by FSL slicer when making slice images). I can't download this template from that repository. What do you think is best? Perhaps I could download it and then skull strip it? This seems a bit much though...

cmaumet replied:
How about either 1) using the non-skull-stripped version or 2) doing the skull-stripping after download (only once)?

cmaumet commented on nidmviewerfsl/lib/pageGen.py:
```diff
@@ -450,20 +450,35 @@ def generatePostStatsHTML(graph, postStatsFilePath="postStats.html"):
-                            excursionSetSliceImage[i])
-                        ).decode())
+
+                # If the slice image already exists add it.
```
Suggestion of edit to make the code more generic: 
 1. Always check whether the slice image is present (regardless of software) and if absent recreate the slice view.
 2. Use a variable to store the software name and then use the same code for all software if the slice view is missing

?

cmaumet commented on nidmviewerfsl/lib/slicerTools.py:
```diff
@@ -1,195 +1,87 @@
-    # Open the matrix file.
-    matrixFile = open(os.path.join(tempDir, 'resizeMatrix.mat'), 'w')
+    # Resample if necessary
+    if nifDim(exc_set, 'x') > nifDim(template, 'x'):
```
Can we check the need for reslicing using the complete affine matrix (cf. ttp://nipy.org/nibabel/nifti_images.html#the-nifti-affines)? This will ensure that two images with the same size but different orientation get resliced.

TomMaullin replied:
I am a little confused by this. Are you suggesting we check if the two affine matrices are identical in order to decide whether or not to resample?

cmaumet replied:
Yes, that's right! Do you think there is something wrong with using the affine matrices?

cmaumet commented on nidmviewerfsl/lib/slicerTools.py:
```diff
@@ -210,38 +102,63 @@ def getSliceImageFromNifti(tempDir, outputName):
+            'templates', 'T1_skullStripped.nii')
+
+    if nifDim(exc_set, 'x') != nifDim(template, 'x'):
+        # Check which is bigger and resize if necessary
```
Can you briefly explain why you are reslicing one way around or the other around depending on size?

TomMaullin replied:
When I first coded this, resampling the smaller image to the larger image sometimes affected the image in strange ways. Blank horizontal lines appeared and made the image appear sketchy.

cmaumet replied:
Okay... Let's be consistent. Can you remove this condition on size and if we have the black horizontal lines we can investigate why this is happening? (We might have to adapt the order of the spline interpolation, i.e. probably use 0 for nearest neighbors).

cmaumet reviewed (commented):
Hi @TomMaullin,

This looks good to me! I only have a couple of small comments.  

My only worry is about including the NIfTI files in the GitHub repo... Let's discuss directly in the comments.

Thanks!!

TomMaullin committed:
code restructure + horrible bug fix

TomMaullin committed:
PEP

cmaumet commented:
Hi @TomMaullin! Sorry for the delay. I have rereplied to your comments above :)

TomMaullin committed:
resizing only excursion set now, using affines

TomMaullin commented:
Hi @cmaumet , I have addressed most of the feedback you have given... the only exception being the SPM template which I am still not sure on... I have replied above!

TomMaullin committed:
PEP8

TomMaullin commented:
Hi @cmaumet , I have made all requested changes apart from the SPM template issue, which I have commented on again above!

cmaumet commented:
Thanks @TomMaullin! I've commented above.

TomMaullin committed:
Templates removed

TomMaullin committed:
git ignore updated

TomMaullin committed:
tests updated

TomMaullin committed:
PEP8

TomMaullin commented:
Hi @cmaumet , all feedback is now addressed! I have changed it so that the template is downloaded from that repository as suggested! Please let me know if you have any further feedback/comments!

cmaumet commented:
Looks good! Thanks! +1 to merge
